// hyperloop-metabase/lib/generate/class.js
/**
 * Hyperloop Metabase Generator
 * Copyright (c) 2015 by Appcelerator, Inc.
 */
var util = require('./util');

function makeClass (json, cls) {
	var entry = {
		class: {
			name: cls.name,
			properties: [],
			instance_methods: [],
			class_methods: []
		},
		framework: cls.framework,
		filename: cls.filename,
		imports: {},
		superclass: cls.superclass && json.classes[cls.superclass]
	};
	cls.properties && Object.keys(cls.properties).sort().forEach(function (k) {
		var prop = util.generateProp(entry, json, cls.properties[k]);
		entry.class.properties.push(prop);
	});
	cls.methods && Object.keys(cls.methods).sort().forEach(function (k) {
		var method = cls.methods[k];
		if (!method.framework) {
			method.framework = cls.framework;
		}
		if (cls.properties && k in cls.properties) {
			return;
		}
		if (method.instance) {
			entry.class.instance_methods.push(util.generateInstanceMethod(entry, json, method));
		} else {
			entry.class.class_methods.push(util.generateClassMethod(entry, json, method));
		}
	});
	entry.imports = util.makeImports(json, entry.imports);
	return entry;
}

/**
 * generate a class file
 */
function generate (dir, json, cls) {
	var output = util.generateTemplate('class', {
		data: makeClass(json, cls)
	});
	util.generateFile(dir, 'class', cls, output);
}

exports.generate = generate;
